### ISSUE TITLE
A starter .perltidyrc

### ISSUE THREAD

**Issue by swelljoe** [webmin/webmin#650]
I've been using this .perltidyrc in my projects (borrowed from, I think, the Mojolicious project) with some minor modifications for a few months now, and it works for me.

I'd love to see Webmin standardize on a style and automatically apply it across everything. It doesn't have to be *this* configuration; any configuration is fine as long as it's standard across the board and I can apply it to my own modificatons automatically.

The obvious potential point of contention is the 2 column indentation. I don't really care what we choose (though 8 seems large; 4 or 2 are fine with me). 4 is probably closer to what you're used to, so that may be the better option. But, really, I'll accept any style you like, as long as it's consistently applied (and I can auto-apply it to my own files).

```perl
-pbp     # Start with Perl Best Practices
-nst     # Non-standard output (allow operate in-place)
-w       # Show all warnings
-iob     # Ignore old breakpoints
-l=80    # 80 characters per line
-mbl=2   # No more than 2 blank lines
-i=2     # Indentation is 2 columns
-ci=2    # Continuation indentation is 2 columns
-vt=0    # Less vertical tightness
-pt=2    # High parenthesis tightness
-bt=2    # High brace tightness
-sbt=2   # High square bracket tightness
-isbc # Don't indent comments without leading space
```

To test it to see what it produces, you can put it into a file called .perltidyrc in the project directory, or you can use the `-pro=filename` option to specify where to load the configuration from. By default, it'll create a new file called whatever.tdy rather than replacing the file.

**Comment by qooob** [comment]
Should I also format theme's files as such as well?

**Comment by qooob** [comment]
I actually already did, just with slightly other configuration.

**Comment by swelljoe** [comment]
Yes, and I'd love it if we could all agree on *one* configuration for use across everything (I'll reformat mine if it's different from what I'm currently using). What changes did you make?

I don't have strong opinions on most of the details, I just want a consistent style across the whole project, if we can all agree on something.

**Comment by qooob** [comment]
Joe, that would be great if you'd like my _.perltidyrc_.

It's important for me not to break each concatenated piece of code, as well as have more than 80 chars in the line. There are few more changes. Have a look:

```Bash
-lp
-vt=1
-sbl
-wba='or'
-vtc=1
-l=125
-i=4
-ce
-pt=2
-sbt=2
-bbt=2
-nsfs
-perl
-sbt=1
-baao
```

**Comment by jcameron** [comment]
I do have a style I use in the core Webmin code, although it's not encoded anywhere (like in a perltidyrc)

**Comment by swelljoe** [comment]
OK, I'll try to figure out how to replicate that. I guess an actively maintained Virtualmin module, maybe virtualmin-nginx, would be a good place to look for what that style looks like?

It looks like we've got tabs (instead of spaces). What tab length do you use in your editor, Jamie? It'll make a difference in how perltidy aligns things that are intended to be aligned, e.g.:

my @tds = ( 'stuff', 'things',
                    'more, 'other' );

Assuming that's the style we want. It looks like they are intended to align, in cases like this (though it doesn't on my system, as I have 2 space tabs configured in vim):

```perl
print &ui_buttons_row("restart.cgi", $text{'index_restart'},
            $text{'index_restartdesc'});
```

Looks like it aligns with tabstop=8, so I'll have to crush down my feelings and accept 8 char tabs. ;-)

Indented closing braces needs to combine several options to replicate, and I think `-otr` will do it.

That will, I think, do curlies like this:

```perl
if (expr) {
  ...
  }
```

I'll go poke at it some more, post an updated perltidyrc once I have something that mostly leaves recent Webmin code unmodified.

**Comment by jcameron** [comment]
Yes, I use 8-space tabs. The style I use for `if` statements are like : 

```
if (foo) {
        something
} else {
        blah
}
```

**Comment by swelljoe** [comment]
That's inconsistent with what I'm seeing for if/while blocks, and maybe all other blocks except subroutine blocks.

e.g., from virtualmin-nginx-lib.pl:

```perl
if (!$get_config_cache) {
        $get_config_cache = &read_config_file($config{'nginx_config'});
        }
```

But, I much prefer the style you've shown above that aligns the closing curly with the opening `if` (to me, it's easier to recognize when the block ends and at what depth I'm reading). Are you OK with that style for all blocks?

I think it'll change subroutine blocks, as well, which are different from other blocks. e.g. current code seems to be like this:

```perl
sub name
{
...
}
```

So, it'd end up becoming:

```perl
sub name {
    ...
}
```

Changing both the opening curly location and the indentation of the function body. I'm very much in favor of a consistent block style across all block usages (e.g. sub, if, while, etc.). But, it seems like some usages will need to change (automatically, nobody will have to do it manually).

It feels kinda like the non-indented subs is to mitigate how far right things go with 8 char tabstops. Maybe 4 char tabstops would make it feel more comfortable to indent subs. ;-)

Anyway, perltidy can completely automate whatever changes we want, and apply it consistently across the whole codebase (and I'll write something to handle it whenever we want to call it). I just need to sort out what you'd like it to look like.

Maybe I'm just looking at a bad example that's mostly of an older style than what you're doing now. Is there something newer than virtualmin-nginx worth looking at to sort out the style you like?

**Comment by jcameron** [comment]
Damn, I was a victim of my own auto-formatter! The correct format is like : 

```
if (foo) {
        something
        }
else {
        blah
        }
```

I admit I violate this for subs to avoid a space-wasting unnecessary indent.

**Comment by shoulders** [comment]
this milestone for webmin needs to be deleted 😄 **2017**

**Comment by iliaross** [comment]
I believe we should all adopt Jamie’s variation of Ratliff indentation style. It would make much more sense, considering he has written more lines of code than anyone else.

I've used it with all the new code I write for Webmin and modules, and I like it. It's fine!

**Comment by jcameron** [comment]
> I believe we should all adopt Jamie’s variation of Ratliff indentation style. It would make much more sense, considering he has written more lines of code than anyone else.

Yes please do!

**Comment by swelljoe** [comment]
I've got a `.perltidyrc` that respects pretty much everything in Webmin, _except_ the lack of indentation in subroutines (i.e. the first level of a subroutine is at 0), e.g.:

```
# get_config_files()
# Returns a list of all firewalld config files
sub get_config_files
{
my $conf_dir = $config{'config_dir'} || '/etc/firewalld';
return (glob("$conf_dir/*.xml"), glob("$conf_dir/*/*.xml"));
}
```

`perltidy` has no option for this convention.

The `.perltidyrc`:

```
# Webmin perltidy configuration
#
# Key characteristics:
#   - 8-column hard tabs for indentation
#   - Ratliff/Banner brace style (closing brace at body indentation)
#   - Sub opening brace on its own line
#   - Uncuddled else/elsif
#   - Logical operators (|| &&) at end of line, not start of continuation
#   - No spaces around string concatenation operator (.)
#
# Known limitations:
#   1. The legacy Webmin convention of not indenting code inside sub
#      bodies (body at the same column as the sub keyword) is not
#      expressible in perltidy. Running perltidy on existing code will
#      add one level of indentation to all sub body content. This is
#      the most significant formatting change vs. existing code.
#   2. Deeply nested UI helper calls (ui_table_row, ui_select, etc.)
#      may reflow differently than the hand-formatted originals. The
#      -lp (line-up-parens) option was considered but produces worse
#      results with Webmin's deeply nested call patterns.

# Indentation
-i=8            # 8 columns per indent level
-ci=4           # Continuation indent (half of indent level)
-et=8           # Entab leading whitespace, 8-column tab stops

# Line length
-l=100          # Max line length

# Brace placement
-nbl            # Control-structure opening brace stays on same line
-sbl            # Sub opening brace goes on a new line by itself
-icb            # Indent closing brace to body level (Ratliff/Banner)

# else / elsif
-nce            # Uncuddled: else/elsif on its own line, not after }

# Container tightness
-pt=2           # Parens:   f($x)    not  f( $x )
-sbt=2          # Brackets: $a[0]    not  $a[ 0 ]
-bt=2           # Braces:   $h{key}  not  $h{ key }

# Blank lines
-bbs            # Blank line before subs
-mbl=2          # Max 2 consecutive blank lines
-kbl=1          # Keep existing blank lines

# Spacing
-nwls="."       # No space to the left of . (concat)
-nwrs="."       # No space to the right of . (concat)

# Line break positions
-wba="&& || and or ."  # Break after these operators (keep at end of line)

# Other
-nolq           # Don't outdent long quoted strings
```

Given that there's no way to make perltidy respect the current style, I guess we should just make it available for use in third-party modules that want to mostly adhere to Webmin style.

**Comment by shoulders** [comment]
maybe add this file to the docs for developers.

I don't know if these recommendations are listed anywhere already.

**Comment by iliaross** [comment]
> Given that there's no way to make perltidy respect the current style, I guess we should just make it available for use in third-party modules that want to mostly adhere to Webmin style.

Thanks, Joe! That’s awesome. I’ve been using your `.perltidyrc` (with a few tweaks) and turned it into a script that pretty much nails a Jamie-style take on the Ratliff format, with proper function indentation and a few extra wrapping adjustments.

See this [`perltidy.pl`](https://github.com/webmin/webmin/blob/master/perltidy.pl), which is now included in the Webmin repository.

You can use it as simply as:

```
./perltidy.pl /path/to/perl/file/to/format
```

**Comment by swelljoe** [comment]
That's a lot of code. But, I guess it if does the thing reliably. 🤷‍♂️

**Comment by iliaross** [comment]
Well, yes, there is some code, as I don't think there's another way to solve it.

**Comment by iliaross** [comment]
I've labeled  it as "Needs Work" so we can review it further without it being flagged for cleanup.

Yet, if you think it's already resolved or working as intended, go ahead and close it now.

**Comment by iliaross** [comment]
I think we’ve reached a point where, if we want a perfect one-to-one match with Jamie’s syntax, we’ll need to use AI for that.

I can handle that if you’d like me to. However, the current version already works well enough, so keeping it simpler might be the better option.

And, if so, we should close this one.

**Comment by iliaross** [comment]
> That's a lot of code. But, I guess it if does the thing reliably. 🤷‍♂️

I’ve reviewed this issue, and I think we can close it as solved since we already have `.perltidyrc` as well as a custom script that makes the formatting for new files very similar to the syntax Jamie uses.

**Comment by shoulders** [comment]
stupid questions, where is the `.perltidyrc` I thought it should be in the root of webmin repo for use by IDE's etc..

**Comment by iliaross** [comment]
It’s now all part of the [perltidy.pl](https://github.com/webmin/webmin/blob/master/perltidy.pl) script, as `.perltidyrc` by itself won’t properly fix the issue.
